app/cosi/main.go

```diff
@@ -0,0 +1,253 @@
+	cosi -m “<Message to be signed>” -c <cosi-group.toml>
+If you would instead like to sign a message contained in a file you can the following command:
+	cosi -f <file-to-be-signed> -c <cosi-group.toml>
+It will create a file file-to-be-signed.sig containing the hash of the file and the signature.
+
+To verify the signature on a file add the verify flag (-v):
+	cosi -f <file-to-be-signed> -c <cosi-group.toml> -v
+This command opens the corresnponding .sig file and validates the contained signature.
+
+Example usuage (create a file my-local-group.toml using cosid first):
+cosi -f my-file.txt -c my-local-group.toml`)
+	os.Exit(1)
+}
+
+var f *flag.FlagSet
+var m *flag.FlagSet
+
+func init() {
+	f = flag.NewFlagSet("f", flag.ContinueOnError)
+	m = flag.NewFlagSet("m", flag.ContinueOnError)
+	//dbg.SetDebugVisible(3)
+}
+
+func main() {
+	lenArgs := len(os.Args)
+	if !(5 <= lenArgs && lenArgs <= 6) {
+		printUsageAndExit("Not enough arguments provided.\n")
+	}
+	switch os.Args[1] {
+	case "-f":
+		strOrFilename := f.String("f", "",
+			"Filename of the file to be signed.")
+		groupToml := f.String("c", "",
+			"Toml file containing the list of CoSi nodes.")
+		verify := f.Bool("v", false, "Verify the files signature.")
+		if err := f.Parse(os.Args[1:]); err != nil {
+			printUsageAndExit("Unable to start signing file. " +
+				"Couldn't parse arguments:" + err.Error())
+		}
+		if *verify {
+			err := verifyFileSig(*strOrFilename, *groupToml)
+			printVerificationResult(err)
+		} else {
+			sig, err := signFile(*strOrFilename, *groupToml)
+			handleErrorAndExit("Couldn't create signature", err)
+			sigFileName := *strOrFilename + ".sig"
+			outFile, err := os.Create(sigFileName)
+			handleErrorAndExit("Couldn't create signature file", err)
+			writeSigAsJSON(sig, outFile)
+			fmt.Println("Signature written to: " + sigFileName)
+		}
+	case "-m":
+		strOrFilename := m.String("m", "", "Message to be signed.")
+		groupToml := m.String("c", "", "Toml file containing the list "+
+			"of CoSi nodes.")
+		if err := m.Parse(os.Args[1:]); err != nil {
+			printUsageAndExit("Unable to start signing message" +
+				"Couldn't parse arguments:" + err.Error())
+		}
+		sig, err := signString(*strOrFilename, *groupToml)
+		handleErrorAndExit("Couldn't create signature", err)
+		writeSigAsJSON(sig, os.Stdout)
+	default:
+		printUsageAndExit("")
+	}
+}
+
+// signStatement can be used to sign the contents passed in the io.Reader
+// (pass an io.File or use an strings.NewReader for strings)
+func signStatement(r io.Reader,
+	el *sda.EntityList,
+	verify bool) (*sda.CosiResponse, error) {
+
+	msgB, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, err
+	}
+	// create a throw-away key pair:
+	kp := config.NewKeyPair(network.Suite)
+
+	// create a throw-away entity with an empty  address:
+	e := network.NewEntity(kp.Public, "")
+	client := network.NewSecureTcpHost(kp.Secret, e)
+	req := &sda.CosiRequest{
+		EntityList: el,
+		Message:    msgB,
```

Comment:
For the moment I think it's best to keep it like that. When @Liamsi made the remark about going more on the *stream* side, there had been issues raised up:
* In the case the witnesses don't need to verify the *content* of the message, we can pass directly the hash
* In case the witnesses *want* to verify the content of the message (for example a certificate, to verify that the domain is correct etc), the witnesses will have to get the whole file and not only the hash. That is when the streaming ability would be a real asset here. And I guess that will be the most used case.